File: impl/typescript/src/parse.ts
/**
 * Document ↔ model conversion plus semantic validation (SPEC.md §6.2).
 *
 * Stage 1 (structural) is Zod (schema.ts). Stage 2 (semantic) lives here: weight/alignment rules,
 * parameter sanity (via distribution construction), and capability re-validation — the declared
 * capabilities MUST equal the values recomputed from structure, or the document is rejected.
 */

// Core
import { z } from 'zod'
// Services
import { capabilities } from './operations'
import { createDistribution } from './distributions'
import { zDocument, type RawNode } from './schema'
// Types
import {
  capabilitiesEqual,
  type Capabilities,
  type Joint,
  type Leaf,
  type Mixture,
  type RVNode,
  type Support,
  type Transform,
} from './model'
import { buildOp } from './ops'
// Errors
import { CapabilityMismatch, ValidationError } from './errors'

const WEIGHT_TOL = 1e-9

export interface ParseOptions {
  /** Run semantic validation (weights, capability match) after structural parse. Default true. */
  validate?: boolean
}

/** Parse a document object into the RV model, running structural + (by default) semantic validation. */
export function parseDocument(doc: unknown, options: ParseOptions = {}): RVNode {
  const { validate = true } = options
  const parsed = zDocument.safeParse(doc)
  if (!parsed.success) {
    throw new ValidationError(`document failed structural validation: ${formatZod(parsed.error)}`)
  }
  const node = parseNode(parsed.data.rv)
  if (validate) validateSemantics(node)
  return node
}

/** Convert a structurally-valid raw node into the rich model (builds Strategy ops, etc.). */
export function parseNode(raw: RawNode): RVNode {
  switch (raw.kind) {
    case 'leaf':
      return {
        kind: 'leaf',
        dist: raw.dist,
        params: raw.params,
        ...(raw.support
          ? {
              support: {
                ...(raw.support.lower !== undefined ? { lower: raw.support.lower } : {}),
                ...(raw.support.upper !== undefined ? { upper: raw.support.upper } : {}),
                lower_inclusive: raw.support.lower_inclusive ?? true,
                upper_inclusive: raw.support.upper_inclusive ?? true,
              },
            }
          : {}),
        ...(raw.capabilities ? { declared: raw.capabilities } : {}),
      }
    case 'joint':
      return {
        kind: 'joint',
        dims: raw.dims.map(parseNode),
        ...(raw.capabilities ? { declared: raw.capabilities } : {}),
      }
    case 'mixture':
      return {
        kind: 'mixture',
        weights: raw.weights,
        components: raw.components.map(parseNode),
        ...(raw.capabilities ? { declared: raw.capabilities } : {}),
      }
    case 'transform':
      return {
        kind: 'transform',
        base: parseNode(raw.base),
        op: buildOp(raw.op.name, raw.op.params),
        ...(raw.capabilities ? { declared: raw.capabilities } : {}),
      }
  }
}

/** Recursively enforce semantic invariants and capability consistency (SPEC.md §6.2). */
export function validateSemantics(node: RVNode): void {
  switch (node.kind) {
    case 'leaf': {
      // Constructing the distribution validates parameters (and, for empirical, the bulk_ref).
      createDistribution(node.dist, node.params)
      if (node.dist === 'categorical') {
        const probs = node.params['probs'] as number[]
        const categories = node.params['categories'] as number[]
        checkWeights(probs, 'categorical probs')
        if (categories.length !== probs.length) {
          throw new ValidationError('categorical categories/probs length mismatch')
        }
      }
      break
    }
    case 'joint':
      node.dims.forEach(validateSemantics)
      break
    case 'mixture':
      if (node.weights.length !== node.components.length) {
        throw new ValidationError('mixture weights/components length mismatch')
      }
      checkWeights(node.weights, 'mixture weights')
      node.components.forEach(validateSemantics)
      break
    case 'transform':
      validateSemantics(node.base)
      break
  }

  if (node.declared) {
    const computed = capabilities(node)
    if (!capabilitiesEqual(node.declared, computed)) {
      throw new CapabilityMismatch(
        `declared capabilities ${JSON.stringify(node.declared)} != computed ${JSON.stringify(computed)}`,
      )
    }
  }
}

/** Serialize a model node back to a plain document object, emitting recomputed capabilities. */
export function toDict(node: RVNode): Record<string, unknown> {
  const caps: Capabilities = capabilities(node)
  switch (node.kind) {
    case 'leaf': {
      const out: Record<string, unknown> = {
        kind: 'leaf',
        dist: node.dist,
        params: node.params,
        capabilities: caps,
      }
      if (node.support) out['support'] = supportDict(node.support)
      return out
    }
    case 'joint':
      return { kind: 'joint', dims: node.dims.map(toDict), capabilities: caps }
    case 'mixture':
      return {
        kind: 'mixture',
        weights: node.weights,
        components: node.components.map(toDict),
        capabilities: caps,
      }
    case 'transform':
      return {
        kind: 'transform',
        base: toDict(node.base),
        op: node.op.toJSON(),
        capabilities: caps,
      }
  }
}

export interface ToDocumentOptions {
  format_version?: string
  metadata?: Record<string, unknown>
}

export function toDocument(node: RVNode, options: ToDocumentOptions = {}): Record<string, unknown> {
  const { format_version = '1.0.0', metadata } = options
  const doc: Record<string, unknown> = { format_version, rv: toDict(node) }
  if (metadata !== undefined) doc['metadata'] = metadata
  return doc
}

function supportDict(s: Support): Record<string, unknown> {
  const out: Record<string, unknown> = {
    lower_inclusive: s.lower_inclusive,
    upper_inclusive: s.upper_inclusive,
  }
  if (s.lower !== undefined) out['lower'] = s.lower
  if (s.upper !== undefined) out['upper'] = s.upper
  return out
}

function checkWeights(weights: readonly number[], label: string): void {
  if (weights.some((w) => w < 0)) throw new ValidationError(`${label} must be non-negative`)
  const sum = weights.reduce((a, b) => a + b, 0)
  if (Math.abs(sum - 1) > WEIGHT_TOL) {
    throw new ValidationError(`${label} must sum to 1 (got ${sum})`)
  }
}

function formatZod(error: z.ZodError): string {
  return error.issues
    .slice(0, 3)
    .map((i) => `${i.path.join('.') || '<root>'}: ${i.message}`)
    .join('; ')
}

// Re-export types referenced by consumers building nodes directly.
export type { Leaf, Joint, Mixture, Transform }

File: impl/rust/src/parse.rs
//! Document ↔ model conversion plus semantic validation (SPEC.md §6.2).
//!
//! serde deserializes the wire document into a `Raw*` representation (structural stage); this module
//! then builds the model and runs the semantic stage: weight/alignment rules, parameter sanity (via
//! distribution construction), and capability re-validation — declared MUST equal recomputed.

use crate::distributions::create;
use crate::errors::{Result, RvError};
use crate::model::{Capabilities, RvNode, Support};
use crate::operations::capabilities;
use crate::ops::Op;
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::HashMap;

const WEIGHT_TOL: f64 = 1e-9;

#[derive(Deserialize)]
struct RawDocument {
    #[allow(dead_code)]
    format_version: String,
    rv: RawNode,
}

#[derive(Deserialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
enum RawNode {
    Leaf {
        dist: String,
        params: Value,
        #[serde(default)]
        capabilities: Option<Capabilities>,
        #[serde(default)]
        support: Option<RawSupport>,
    },
    Joint {
        dims: Vec<RawNode>,
        #[serde(default)]
        capabilities: Option<Capabilities>,
    },
    Mixture {
        weights: Vec<f64>,
        components: Vec<RawNode>,
        #[serde(default)]
        capabilities: Option<Capabilities>,
    },
    Transform {
        base: Box<RawNode>,
        op: RawOp,
        #[serde(default)]
        capabilities: Option<Capabilities>,
    },
}

#[derive(Deserialize)]
struct RawOp {
    name: String,
    #[serde(default)]
    params: Option<HashMap<String, f64>>,
}

#[derive(Deserialize)]
struct RawSupport {
    #[serde(default)]
    lower: Option<f64>,
    #[serde(default)]
    upper: Option<f64>,
    #[serde(default)]
    lower_inclusive: Option<bool>,
    #[serde(default)]
    upper_inclusive: Option<bool>,
}

/// Parse a JSON document string into the RV model, running structural + semantic validation.
pub fn parse_str(json: &str) -> Result<RvNode> {
    let doc: RawDocument =
        serde_json::from_str(json).map_err(|e| RvError::Validation(format!("invalid document: {e}")))?;
    finish(doc)
}

/// Parse a `serde_json::Value` document into the RV model.
pub fn parse_value(doc: &Value) -> Result<RvNode> {
    let doc: RawDocument = serde_json::from_value(doc.clone())
        .map_err(|e| RvError::Validation(format!("invalid document: {e}")))?;
    finish(doc)
}

fn finish(doc: RawDocument) -> Result<RvNode> {
    let node = build(doc.rv)?;
    validate_semantics(&node)?;
    Ok(node)
}

fn build(raw: RawNode) -> Result<RvNode> {
    match raw {
        RawNode::Leaf { dist, params, capabilities, support } => Ok(RvNode::Leaf {
            dist,
            params,
            support: support.map(build_support),
            declared: capabilities,
        }),
        RawNode::Joint { dims, capabilities } => Ok(RvNode::Joint {
            dims: dims.into_iter().map(build).collect::<Result<_>>()?,
            declared: capabilities,
        }),
        RawNode::Mixture { weights, components, capabilities } => Ok(RvNode::Mixture {
            weights,
            components: components.into_iter().map(build).collect::<Result<_>>()?,
            declared: capabilities,
        }),
        RawNode::Transform { base, op, capabilities } => Ok(RvNode::Transform {
            base: Box::new(build(*base)?),
            op: Op::build(&op.name, &op.params.unwrap_or_default())?,
            declared: capabilities,
        }),
    }
}

fn build_support(rs: RawSupport) -> Support {
    Support {
        lower: rs.lower,
        upper: rs.upper,
        lower_inclusive: rs.lower_inclusive.unwrap_or(true),
        upper_inclusive: rs.upper_inclusive.unwrap_or(true),
    }
}

/// Recursively enforce semantic invariants and capability consistency (SPEC.md §6.2).
pub fn validate_semantics(node: &RvNode) -> Result<()> {
    match node {
        RvNode::Leaf { dist, params, .. } => {
            // Constructing the distribution validates parameters (and, for empirical, the bulk_ref).
            create(dist, params)?;
            if dist == "categorical" {
                let probs = f64_array(params, "probs")?;
                let categories = f64_array(params, "categories")?;
                check_weights(&probs, "categorical probs")?;
                if categories.len() != probs.len() {
                    return Err(RvError::Validation("categorical categories/probs length mismatch".into()));
                }
            }
        }
        RvNode::Joint { dims, .. } => {
            for dim in dims {
                validate_semantics(dim)?;
            }
        }
        RvNode::Mixture { weights, components, .. } => {
            if weights.len() != components.len() {
                return Err(RvError::Validation("mixture weights/components length mismatch".into()));
            }
            check_weights(weights, "mixture weights")?;
            for comp in components {
                validate_semantics(comp)?;
            }
        }
        RvNode::Transform { base, .. } => validate_semantics(base)?,
    }

    if let Some(declared) = node.declared() {
        let computed = capabilities(node)?;
        if declared != computed {
            return Err(RvError::CapabilityMismatch(format!(
                "declared {declared:?} != computed {computed:?}"
            )));
        }
    }
    Ok(())
}

/// Serialize a model node back to a wire document object, emitting recomputed capabilities.
pub fn to_value(node: &RvNode) -> Result<Value> {
    let caps = serde_json::to_value(capabilities(node)?).unwrap();
    Ok(match node {
        RvNode::Leaf { dist, params, support, .. } => {
            let mut obj = json!({ "kind": "leaf", "dist": dist, "params": params, "capabilities": caps });
            if let Some(s) = support {
                obj["support"] = support_value(s);
            }
            obj
        }
        RvNode::Joint { dims, .. } => {
            let ds: Vec<Value> = dims.iter().map(to_value).collect::<Result<_>>()?;
            json!({ "kind": "joint", "dims": ds, "capabilities": caps })
        }
        RvNode::Mixture { weights, components, .. } => {
            let cs: Vec<Value> = components.iter().map(to_value).collect::<Result<_>>()?;
            json!({ "kind": "mixture", "weights": weights, "components": cs, "capabilities": caps })
        }
        RvNode::Transform { base, op, .. } => {
            json!({ "kind": "transform", "base": to_value(base)?, "op": op.to_json(), "capabilities": caps })
        }
    })
}

/// Serialize a full document (`format_version` + optional `metadata` + `rv`).
pub fn to_document(node: &RvNode, format_version: &str, metadata: Option<Value>) -> Result<Value> {
    let mut doc = json!({ "format_version": format_version, "rv": to_value(node)? });
    if let Some(m) = metadata {
        doc["metadata"] = m;
    }
    Ok(doc)
}

fn support_value(s: &Support) -> Value {
    let mut obj = json!({ "lower_inclusive": s.lower_inclusive, "upper_inclusive": s.upper_inclusive });
    if let Some(l) = s.lower {
        obj["lower"] = json!(l);
    }
    if let Some(u) = s.upper {
        obj["upper"] = json!(u);
    }
    obj
}

fn check_weights(weights: &[f64], label: &str) -> Result<()> {
    if weights.iter().any(|&w| w < 0.0) {
        return Err(RvError::Validation(format!("{label} must be non-negative")));
    }
    let sum: f64 = weights.iter().sum();
    if (sum - 1.0).abs() > WEIGHT_TOL {
        return Err(RvError::Validation(format!("{label} must sum to 1 (got {sum})")));
    }
    Ok(())
}

fn f64_array(params: &Value, key: &str) -> Result<Vec<f64>> {
    params
        .get(key)
        .and_then(Value::as_array)
        .map(|a| a.iter().filter_map(Value::as_f64).collect())
        .ok_or_else(|| RvError::Validation(format!("expected numeric array '{key}'")))
}
